Tests: Cover Block Hooks integration with a custom block theme

gziolo opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/59313

For start, I created a simplified version of [Twenty Twenty-Three theme](https://github.com/WordPress/wordpress-develop/tree/trunk/src/wp-content/themes/twentytwentythree) that we used for testing Block Hooks integration that contains:

-  The required `index.html` and the `single.html` template that I plan to use with tests.
- 3 template oarts where two of them reference patterns.
- 3 Patterns referenced in templates and template parts.

I want to automatically register 4 custom blocks where each of them hooks into another block using all four target relative positions: `before`, `after`, `firstChild`, `lastChild`.

Now, the goal would be to verify that the block gets hooked into the correct positions when targeting:
- template
- template part
- pattern

Bonus points would be for using a filter to prevent insertion of the block based on some made-up conditions.

Separately from that, with such a block theme in place and four hooked blocks, we could run some profiling on the block pattern registry that will contain at least 3 patterns to see if `get_hooked_blocks` is really the hot path we should be worried about.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

gziolo committed:
Tests: Cover Block Hooks integration with a custom block theme

gziolo committed:
Fix the test listing test themes

gziolo committed:
Improve hooked blocks integration wit the theme

gziolo committed:
Add tests covering loading template and template part

gziolo committed:
Align spaces for assignments

gziolo committed:
Add test covering block patterns

gziolo commented:
> Bonus points would be for using a filter to prevent insertion of the block based on some made-up conditions.
>
> Separately from that, with such a block theme in place and four hooked blocks, we could run some profiling on the block pattern registry that will contain at least 3 patterns to see if get_hooked_blocks is really the hot path we should be worried about.

This PR is big enough, so I would prefer to split other test scenarios into a follow-up work.

gziolo committed:
Fix spaces used instead of tabs

ockham commented on tests/phpunit/tests/blocks/blockHooks.php:
```diff
@@ -8,26 +8,76 @@
+	/**
+	 * @see register_block_type()
+	 */
+	private function register_block_type( $block_type, $args = array() ) {
+		$result                         = register_block_type( $block_type, $args );
+		$this->registered_block_names[] = $result->name;
+
+		return $result;
+	}
```
One alternative to "shadowing" `register_block_type` that I've seen elsewhere is using a private instance of the block type registry ([e.g.](https://github.com/WordPress/wordpress-develop/blob/f7d73158acd8bef9c0d98316c1cd5ecf4cd1a636/tests/phpunit/tests/blocks/wpBlockTypeRegistry.php#L13-L19)). (That might make setup and teardown a bit cleaner, and avoid helper member variables such as `registered_block_names`.)

gziolo replied:
I shamelessly stole this code from the test suite covering Block Context:

https://github.com/WordPress/wordpress-develop/blob/f7d73158acd8bef9c0d98316c1cd5ecf4cd1a636/tests/phpunit/tests/blocks/context.php#L37-L47

https://github.com/WordPress/wordpress-develop/blob/f7d73158acd8bef9c0d98316c1cd5ecf4cd1a636/tests/phpunit/tests/blocks/context.php#L49-L66

These blocks need to be in the global registry from `get_hooked_blocks` to work correctly, and I only added the logic to track registered blocks to unregister them later.

A simpler alternative might be to iterate over all registered blocks and deregister these starting with `tests/` 🤔 What do you think? It could also work with patterns starting with the name of the test theme.

gziolo replied:
I tied a revised approach with https://github.com/WordPress/wordpress-develop/pull/5340/commits/cd6f58204d2bbde67804f7eded0fe776b4a50bed. It's all contained in the `tear_down` now. If you have ideas on how to improve it further, let me know.

ockham commented on tests/phpunit/data/themedir1/block-theme-hooked-blocks/patterns/hidden-comments.php:
I like that this is a realistic example of a pattern, but OTOH, it's a bit big and complex, so it's not entirely easy to know what to look out for (i.e. where blocks are going to be inserted). Maybe a slightly more minimal pattern might make sense?

🔢 Applies elsewhere (to the post meta pattern, and the theme's index and single templates).

gziolo replied:
I have already removed a ton of less essential attributes from the block metadata and HTML elements that were present in the Twenty Twenty-Three theme. I see your point, and I'm happy to see what else can be removed from the files. Do you have any recommendations that could make it easier to follow?

One important note that might be unclear at the moment. I kind of like the complexity in light of the planned effort to run some profiling on `do_blocks` call that renders a single post page using some test post data. It might allow us to confirm whether `get_hooked_blocks` is performant enough and what is the impact of the filters implemented on top of it.

ockham commented on tests/phpunit/tests/blocks/blockHooks.php:
```diff
@@ -119,4 +169,98 @@ public function test_get_hooked_blocks_matches_found() {
+			'<!-- wp:post-content {"layout":{"type":"constrained"}} /-->'
+			. '<!-- wp:tests/hooked-after /-->',
+			$template->content
+		);
```
Nice 👍

ockham commented on tests/phpunit/tests/theme/themeDir.php:
```diff
@@ -185,6 +185,7 @@ public function test_theme_list() {
 			'Block Theme [0.4.0]',
 			'Block Theme [1.0.0] in subdirectory',
 			'Block Theme Deprecated Path',
+			'Block Theme Hooked Blocks',
```
Nitpick, the name is a bit clunky (especially when read as a path name, `block-theme-hooked-blocks`). How about something shorter, e.g. `Hooked Blocks Theme`? 🤔 (Although I realize we might want to have it next to the other `Block Theme ...` test themes, so maybe not a good idea 😅 )

gziolo replied:
I mostly wanted to have the same prefix:

<img width="429" alt="Screenshot 2023-10-02 at 18 02 04" src="https://github.com/WordPress/wordpress-develop/assets/699132/0b4ff2b1-7a66-40e9-851d-72b09e412854">

It could also become `Block theme with hooked blocks`. I'm open to changing it.

gziolo replied:
Trying `block-theme-with-hooked-blocks` and `Block Theme with Hooked Blocks` in https://github.com/WordPress/wordpress-develop/pull/5340/commits/2caf1146e8e7706b166e9325aa86399b1182bd9b.

ockham replied:
Thank you, I like that better 😄

ockham reviewed (approved):
Thanks a lot, it's great to have some proper coverage for some of the most relevant cases we'd like to guard against regressions 👍 

I left a few minor notes, but none of them blocking. Feel free to disregard and merge 😄 If needed, we can tackle them in a follow-up.

gziolo committed:
Rename the test suite file

gziolo committed:
Simplify the test cleanup for block types and patterns

gziolo committed:
Rename the test theme

gziolo commented:
I addressed all comments, but https://github.com/WordPress/wordpress-develop/pull/5340#discussion_r1342899705. As noted, I'm happy to iterate on the test theme and reduce it further after profiling Block Hooks usage in a real-life scenario when rendering a single post page.

gziolo commented:
Committed with https://core.trac.wordpress.org/changeset/56759.
